### PR TITLE
Adding probes to opensearch, opensearch-dashboards

### LINKED ISSUE #44: [Enhancement][opensearch][opensearch-dashboards] Lifecycle Probes
**Is your feature request related to a problem? Please describe.**
I think there ought to be liveness/readiness/startup probes for the opensearch statefulset and the opensearch-dashboards deployment.

Currently the pods are immediately marked ready on startup, leading to more than one of the nodes being unavailable which can lead to instability especially with lots of data.

**Describe the solution you'd like**
I am happy to implement something in a PR to resolve this, if that is something the project would want and my general ideas below sound good.

Opensearch:
Readiness: query health endpoint, which may require whitelisting or auth, could potentially look for green cluster status
Liveness: can check for tcp connection to 9200 or 9300
Startup: can check for health endpoint regardless of status?

Opensearch-dashboards:
Readiness: Health is at least yellow in /api/status
Liveness: 5601 is open to tcp
Startup: same as readiness

**Describe alternatives you've considered**
NA

**Additional context**
I am happy to implement if the ideas sound good.

### PR DESCRIPTION
**Pull request by hgoscenski-imanage**
Initial implementations of probes for opensearch and
opensearch-dashboards.

Default probes are reasonable to use, though for opensearch readiness it will require
credentials.

all probes have `custom` option where the end user can pass in whatever
check they prefer.

Signed-off-by: Harrison Goscenski <harrison.goscenski@imanage.com>

### Description
Adds liveness, readiness, and startup probes
 
### Issues Resolved
#44 
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by hgoscenski-imanage** [comment]
@peterzhuamazon I have updated this with some basic probes, and the option for entire custom probes. The readiness probe is something that works well for how I am utilizing the chart, but I believe leaving it disabled by default is the correct choice as there is underlying security configuration required to allow for the probe to work.

**Comment by peterzhuamazon** [comment]
> @peterzhuamazon I have updated this with some basic probes, and the option for entire custom probes. The readiness probe is something that works well for how I am utilizing the chart, but I believe leaving it disabled by default is the correct choice as there is underlying security configuration required to allow for the probe to work.

Thanks @hgoscenski-imanage this is a rather large PR I would wait for reviews from @DandyDeveloper @TheAlgo.
In the meantime do you have any test results to share?
Thanks.

**Comment by hgoscenski-imanage** [comment]
@peterzhuamazon it is definitely a sizable PR. I combined it with a few of my other PR's in order to get a version that runs with the configuration I needed and it functions as I would expect.

I configured the master/client nodes to use the simple readiness probe, which remained in the Unready state until the master came up and the securityconfig was initialized (in our deployment we reinitialize the securityconfig on each helm depoyment).

Once it was initialized and the monitoring user was available, the client/masters went to ready, the data nodes were configured with the liveness probe to wait for green to ensure only one node is taken down at a time (in combination with the PDB) ensuring stability. I tested this over the course of a few deploys to verify it behaved in that manner.

The other change I made to our deployment is to ensure that each statefulset (client, master, data) are set to use RollingUpdate instead of parallel to ensure they are handled one at a time and that any misconfigured readiness probe does not take down the entire cluster. 

I am unsure what to attach to prove this however, especially since on its own this PR does not actually run given the other PR's contain needed changes.

File: Helm/opensearch-dashboards/templates/deployment.yaml
Comment (DandyDeveloper) [inline Helm/opensearch-dashboards/templates/deployment.yaml:146]: I think enabling/disabling liveness probe (or any probe) shouldn't be an option here. Map the defaults from the `values.yaml` and users can then just overwrite these parts if they choose. 

I.E;

```yaml
readinessProbe:
  httpGet:
    path: /api/status
    port: http
  periodSeconds: 15
  initialDelaySeconds: 30
  failureThreshold: 3
  timeoutSecond: 1   
livenessProbe:
  tcpSocket:
     port: http
  periodSeconds: 15
  initialDelaySeconds: 30
  failureThreshold: 3
  timeoutSecond: 1 
```
We should do this for all of them. 
 
Then we just 
```
{{- with .Values.livenessProbe }}
{{- . | toYaml | nindent X }}
{{- end }}
```

Customizing can get confusing and really users should change the base params. Just maybe timeouts and things.

File: Helm/opensearch-dashboards/values.yaml
Comment (DandyDeveloper) [inline Helm/opensearch-dashboards/values.yaml:86]: See initial comment. 

```suggestion
readinessProbe:
  tcpSocket:
    port: http
  periodSeconds: 15
livenessProbe:
  abc: xyz
startupProbe:
  abc: xyz
  ...
```

File: Helm/opensearch-dashboards/values.yaml
Comment (DandyDeveloper) [inline Helm/opensearch-dashboards/values.yaml:117]: Unnecessary change

File: Helm/opensearch/templates/lifecycle-secret.yaml
Comment (DandyDeveloper) [inline Helm/opensearch/templates/lifecycle-secret.yaml:15]: See previous recommendations. I'd advise removing the enable condition here. As this should always exist.

File: Helm/opensearch/templates/statefulset.yaml
Comment (DandyDeveloper) [inline Helm/opensearch/templates/statefulset.yaml:253]: If we are happy and this always works. We should not let this be conditional. It's helpful and shouldn't really change.

File: Helm/opensearch/values.yaml
Comment (DandyDeveloper) [inline Helm/opensearch/values.yaml:77]: Maybe keep this to 1 line.

File: Helm/opensearch/values.yaml
Comment (DandyDeveloper) [inline Helm/opensearch/values.yaml:112]: 1 line. Unnecessary change.

File: Helm/opensearch/values.yaml
Comment (DandyDeveloper) [inline Helm/opensearch/values.yaml:126]: 1 line. Unnecessary change.

File: Helm/opensearch/values.yaml
Comment (DandyDeveloper) [inline Helm/opensearch/values.yaml:135]: 1 line. Unnecessary change.

File: Helm/opensearch/values.yaml
Comment (DandyDeveloper) [inline Helm/opensearch/values.yaml:188]: 1 line. Unnecessary change.

File: Helm/opensearch/values.yaml
Comment (DandyDeveloper) [inline Helm/opensearch/values.yaml:193]: 1 line. Unnecessary change.

File: Helm/opensearch/values.yaml
Comment (DandyDeveloper) [inline Helm/opensearch/values.yaml:284]: 1 line. Unnecessary change.

File: Helm/opensearch/values.yaml
Comment (DandyDeveloper) [inline Helm/opensearch/values.yaml:350]: See previous comments. This is a good way forward but it could be cleaner and shouldn't really be optional.

**Review by DandyDeveloper — CHANGES_REQUESTED** [review]
Thank you for the PR. 

I like the readinessProbe script, that is solid and the inputs there are good. 

I don't think probes should be customized. If we have a reliable and valid mechanism for readiness, it should be pretty static. 

At most, we should allow intervals/timeouts to be changed. 

ALso, be careful with some changes. Your newlines for vars like `lifecycle: {}` were unnecessary and make the PR less fun to work through. Please undo those changes.

**Comment by hgoscenski-imanage** [comment]
@DandyDeveloper I apologize for the unnecessary changes there, it looks like my formatter was a tad bit overzealous. I will revert them all.

File: Helm/opensearch-dashboards/templates/deployment.yaml
Comment (TheAlgo) [reply Helm/opensearch-dashboards/templates/deployment.yaml:146]: @Dandy What if some users don't want to have some probe? I agree on the mandatory `readiness` though.

File: Helm/opensearch-dashboards/values.yaml
Comment (TheAlgo) [inline Helm/opensearch-dashboards/values.yaml:62]: FYI : This config is going to be changed via #38 . Some instances of `dashboards` will get replaced with `opensearch-dashboards` and `opensearch_dashboards` respectively.

File: Helm/opensearch/values.yaml
Comment (TheAlgo) [inline Helm/opensearch/values.yaml:352]: Can we use the naming conventions of following the camel case here as well like the rest of the chart?

**Review by TheAlgo — CHANGES_REQUESTED** [review]

**Comment by TheAlgo** [comment]
@hgoscenski-imanage Overall the PR looks neat and will be very helpful to the end users. We move one step closer to a robust chart for OepnSearch stack. Thanks a ton 😋

File: Helm/opensearch-dashboards/templates/deployment.yaml
Comment (hgoscenski-imanage) [reply Helm/opensearch-dashboards/templates/deployment.yaml:146]: @DandyDeveloper @TheAlgo 
I think being able to disable the probes is a good thing for the users to have, and by default I think not enabling the readiness is good, especially as it is dependent on their security config. If they don't have a robust way of updating their security config their readiness probe would always be failing, which is a bad experience.

File: Helm/opensearch/templates/statefulset.yaml
Comment (hgoscenski-imanage) [reply Helm/opensearch/templates/statefulset.yaml:253]: @DandyDeveloper if we wanted to leave this always available I suppose we could set the default for the acceptable_statuses to `["red", "yellow", "green"]` though I think having the users opt in to that behavior gives them a greater chance of selecting what makes sense in their environment.

Comment (DandyDeveloper) [reply Helm/opensearch/templates/statefulset.yaml:253]: @hgoscenski-imanage 

Actually, I'm second guessing myself here. Do we even want this in the chart? Nodes will still accept documents and process when the shards for that index become available. The cluster can still be red but accepting new documents. 

If we implement this, it means even an `Initialized` node, with recovered shards would not be receiving requests because the k8s service will no longer be routing traffic to them (Because ES is reporting things as red). 

If other nodes are available this wouldn't be a problem, but if you do have a complete cluster failure, or you want to rollout a cluster upgrade, larger clusters with this could be waiting days until a node becomes available after recovering all shards. 

As a user, what scenario would you want ["yellow", "green"] to be the only accepted readiness states?

**Comment by DandyDeveloper** [comment]
@TheAlgo I'm a bit confused by having conditional probes. Why wouldn't a user want liveness, startup and readiness probes by default? 

@hgoscenski-imanage I'm replied to you on the topic of the readiness. In theory I think this is a great idea, but I'm questioning the real world benefits to this. 

Additional thing to note: If any nodes are ready, it won't matter if another node is not ready because the k8s service will route the document to a ready node, then that node will just pass the documents over to the node responsible (assume this is a not ready node) over the transport SSL which uses a headless service (I think?). So... It effectively circumvents the readiness probe anyway.

**Comment by TheAlgo** [comment]
> @TheAlgo I'm a bit confused by having conditional probes. Why wouldn't a user want liveness, startup and readiness probes by default?

@DandyDeveloper I agree with you completely, we can keep the default as enabled but there should be an option to turn it off . As @hgoscenski-imanage mentioned here 

```I think being able to disable the probes is a good thing for the users to have, and by default I think not enabling the readiness is good, especially as it is dependent on their security config. If they don't have a robust way of updating their security config their readiness probe would always be failing, which is a bad experience.```

**Comment by DandyDeveloper** [comment]
@TheAlgo 

> especially as it is dependent on their security config

Hence my general apprehension on actually adding this readiness. In addition to my other observations for @hgoscenski-imanage . 

We may have to simplify the readiness, unless I'm wrong (which I'd like input on if I'm misunderstanding).

**Comment by TheAlgo** [comment]
> We may have to simplify the readiness, unless I'm wrong (which I'd like input on if I'm misunderstanding).
I vote on simplifying the readiness a bit as well @DandyDeveloper

File: Helm/opensearch/templates/statefulset.yaml
Comment (hgoscenski-imanage) [reply Helm/opensearch/templates/statefulset.yaml:288]: @DandyDeveloper I am currently running it in that configuration in our environment, we want to ensure that in combination with the pod disruption budgets that even during automatic draining of nodes that the cluster always remains in an available state. 

This can slow down the rollout process but ensures availability, if the nodes are returning with the `"index.unassigned.node_left.delayed_timeout"` then there would be no unnecessary shuffling of shards.

Our cluster is simply being used to handle logs from a kubernetes cluster and is a relatively small set of data, but I believe there is value in having the ability to configure this. 

All of our log ingesters write to a deployment of opensearch nodes acting as clients which do not check the status of the cluster but rather simply check for the availability of the node. This allow logs to always be written assuming the cluster is maintaining a status where the indices are available, but does imply the necessity of configuration here, at least in our environment.

File: Helm/opensearch/values.yaml
Comment (hgoscenski-imanage) [reply Helm/opensearch/values.yaml:77]: I found it more readable to bring the line length down, but I can certainly revert.

**Comment by DandyDeveloper** [comment]
@hgoscenski-imanage Got some merge conflicts. 

I haven't had the opportunity to review your response yet RE: Readiness. Sorry.

**Comment by TheAlgo** [comment]
@hgoscenski-imanage Apologies but will it be possible for you to raise this PR in our new repo for [Helm Charts](https://github.com/opensearch-project/helm-charts). Thanks

**Comment by peterzhuamazon** [comment]
> @hgoscenski-imanage Apologies but will it be possible for you to raise this PR in our new repo for [Helm Charts](https://github.com/opensearch-project/helm-charts). Thanks

@hgoscenski-imanage Hi would you mind move it to the helm-charts repo? Thanks.

**Comment by hgoscenski-imanage** [comment]
> > @hgoscenski-imanage Apologies but will it be possible for you to raise this PR in our new repo for [Helm Charts](https://github.com/opensearch-project/helm-charts). Thanks
> 
> @hgoscenski-imanage Hi would you mind move it to the helm-charts repo? Thanks.

My apologies for the delay, I have opened https://github.com/opensearch-project/helm-charts/pull/43 for this.

**Pull request closed without merging**
